### PR TITLE
[WIP, but there's just one thing] - MOAIAnimCurveCustom

### PR DESCRIPTION
**Pull request by aab29**
`MOAIAnimCurve` doesn't have many ease functions.  In fact, they're almost all just variations on `powf(t, rate)`.  That sucks.  :spaghetti: 

We could just add our own ease types in C++, but that's not very flexible.  So here's `MOAIAnimCurveCustom`.  It's just like `MOAIAnimCurve`, but it allows you to define a lua function (using `curve:setCallback(easeFunction)`) that will be used for all interpolation by the curve.  :curly_loop: 

This will let us define a library of lua eases to use with `ease()` in the ActionHelpers.  :smoking: 

I haven't noticed any slowdown compared to using regular `MOAIAnimCurve`, even with a bunch of eases running simultaneously, but that's something that we should be watching out for if we start having performance problems.  :bread: 

I should note that this is all based off my brother's work: https://github.com/IkeBart/moai-dev/commit/2e3071931da3e9ac6460645d60311857be150738

---

The one thing that isn't done here has to do with the fact that I don't remember C++.  I'm getting a compiler warning because apparently I don't know how to do inheritance right.

MOAIAnimCurve.h:46: 'virtual void MOAIAnimCurve::GetValue(MOAIAttrOp&, const MOAIAnimKeySpan&)' was hidden
MOAIAnimCurveCustom.h:29: By 'virtual float MOAIAnimCurveCustom::GetValue(const MOAIAnimKeySpan&)'

I want `MOAIAnimCurveCustom` to be just like `MOAIAnimCurve`, except that when it calls `GetValue(const MOAIAnimKeySpan&)`, I want it to use its own method.  That's all.  Can anyone tell me what I'm doing wrong? :question: 

Also, I had to remove a lot of `const` declarations on functions all over `MOAIAnimCurveBase` and its subclasses in order to get `GetValue(const MOAIAnimKeySpan&)` to reference its own `mCallback`.  Does anyone know of a better way to do this?  :question:

### CODE REVIEW

File: src/moaicore/MOAIAnimCurveCustom.cpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+float MOAIAnimCurveCustom::InterpolateCustom(float x0, float x1, float t, float weight) {
+	float v0 = t; // quicker way of doing: USInterpolate::Curve( USInterpolate::kLinear, t);
+	float v1 = 0.0f; // use custom function
+
+	if (this->mCallback) {
+		MOAILuaStateHandle state = this->mCallback.GetSelf();
+		lua_pushnumber(state, t); // add parameter variable t to stack
+		state.DebugCall(1, 1); // the function should take one argument and return one number
```
Comment (JaviSoto) [inline src/moaicore/MOAIAnimCurveCustom.cpp:56]: Is this only for debug?

File: src/moaicore/MOAIAnimCurveCustom.cpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+		int top = state.GetTop();
+		v1 = state.GetValue < float >( top, 0.0f); // store the return value of the function in v1
```
Comment (JaviSoto) [inline src/moaicore/MOAIAnimCurveCustom.cpp:59]: redundant comment

**Comment by JaviSoto** [comment]
> The one thing that isn't done here has to do with the fact that I don't remember C++. I'm getting a compiler warning because apparently I don't know how to do inheritance right.
> 
> MOAIAnimCurve.h:46: 'virtual void MOAIAnimCurve::GetValue(MOAIAttrOp&, const MOAIAnimKeySpan&)' was hidden
> MOAIAnimCurveCustom.h:29: By 'virtual float MOAIAnimCurveCustom::GetValue(const MOAIAnimKeySpan&)'
> 
> I want MOAIAnimCurveCustom to be just like MOAIAnimCurve, except that when it calls GetValue(const MOAIAnimKeySpan&), I want it to use its own method. That's all. Can anyone tell me what I'm doing wrong? 
> 
> Also, I had to remove a lot of const declarations on functions all over MOAIAnimCurveBase and its subclasses in order to get GetValue(const MOAIAnimKeySpan&) to reference its own mCallback. Does anyone know of a better way to do this?  

@NachoSoto you should check this out.

File: src/moaicore/MOAIAnimCurveCustom.cpp (hunk 1 above)
Comment (aab29) [inline src/moaicore/MOAIAnimCurveCustom.cpp:56]: No, it seems like it's used all over Moai.  I have no idea why it's called "debug".  :confused:

**Comment by NachoSoto** [comment]
@aab29 I can take a look at that whenever you have a minute :)

**Comment by jeiting** [comment]
Took a look at the C++ warning.

Basically, because the function we were overriding had a an overloaded version that wasn't covariant with the subclass's implementation, it prevents the overloaded function from being called from the subclass.

It doesn't seem to cause us any problems, but if we ever do need to use this version of the GetValue method we will run into problems. :shipit:

**Comment by JaviSoto** [comment]
That's why we should remove the warning somehow :-/ A warning is a clear code smell: there's something wrong, even if it doesn't affect you now, it could bite you later...

**Comment by JaviSoto** [comment]
And this is the reason why do have `-Werror` on the main app :)

**Comment by jeiting** [comment]
I disagree, warnings are awesome.

**Pull request merged**